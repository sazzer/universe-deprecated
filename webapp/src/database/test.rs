use lazy_static::lazy_static;
use log::info;
use postgres::{Client, NoTls};
use testcontainers::{clients::Cli, images::postgres::Postgres, Container, Docker};

lazy_static! {
    static ref DOCKER: Cli = { Cli::default() };
}

pub struct TestDatabase<'d> {
    #[allow(dead_code)]
    node: Container<'d, Cli, Postgres>,
    url: String,
}

impl<'d> TestDatabase<'d> {
    pub fn new() -> Self {
        let node = DOCKER.run(Postgres::default());

        let host_port = node.get_host_port(5432).unwrap();
        let url = format!("postgres://postgres:postgres@localhost:{}", host_port);
        info!("Running postgres on {}", url);

        TestDatabase { node, url }
    }

    pub fn client(&self) -> Client {
        Client::connect(&self.url, NoTls).unwrap()
    }
}
