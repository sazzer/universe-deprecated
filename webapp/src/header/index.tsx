import React from 'react';

export const Header: React.FC = () => {
  return (
    <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
      <a className="navbar-brand" href="/">Universe</a>
      <button className="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle Navigation">
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbarSupportedContent">
        <ul className="navbar-nav ml-auto">
          <li className="nav-item">
            <a className="nav-link" href="/login">Login / Register</a>
          </li>
        </ul>
      </div>
    </nav>
  );
};
