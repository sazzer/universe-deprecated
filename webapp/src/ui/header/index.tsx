import React from 'react';
import { useTranslation } from "react-i18next";
import { Link } from 'react-router-dom';

export const Header: React.FC = () => {
  const { t } = useTranslation();

  return (
    <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
      <Link to="/" className="navbar-brand">{t('page.title')}</Link>
      <button className="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="{t('header.toggleNavigation')}">
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbarSupportedContent">
        <ul className="navbar-nav ml-auto">
          <li className="nav-item">
            <Link to="/login" className="nav-link">{t('header.loginMenu.title')}</Link>
          </li>
        </ul>
      </div>
    </nav >
  );
};
