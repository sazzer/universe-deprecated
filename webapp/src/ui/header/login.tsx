import { Link } from "react-router-dom";
import React from "react";
import { useTranslation } from "react-i18next";

/** Header link to allow the user to log in */
export const LoginLink: React.FC = () => {
  const { t } = useTranslation();

  return (
    <li className="nav-item">
      <Link to="/login" className="nav-link">
        {t("header.loginMenu.title")}
      </Link>
    </li>
  );
};
