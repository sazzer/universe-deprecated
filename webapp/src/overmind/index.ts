import { namespaced } from "overmind/config";
import { createHook } from "overmind-react";
import { IConfig } from "overmind";
import * as login from "../login";
import * as authentication from "../authentication";

export const config = namespaced({
  login,
  authentication
});

export const useOvermind = createHook<typeof config>();

declare module "overmind" {
  // tslint:disable:interface-name
  interface Config extends IConfig<typeof config> {}
}
