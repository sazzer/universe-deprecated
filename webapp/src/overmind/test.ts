import { createOvermindMock, OvermindMock, Config } from "overmind";
import { config } from "../overmind";
import { cloneDeep, merge } from "lodash";

/**
 * NestedPartial is stolen from Overmind to represent a part of the Overmind structure
 */
export declare type NestedPartial<T> = T extends Function
  ? T
  : Partial<
      {
        [P in keyof T]: NestedPartial<T[P]>;
      }
    >;

/**
 * Create a test instance of the Overmind to work with
 * @param mockEffects The mock effects to wire in
 * @param initialState The initial state to wire in
 */
export function createTestOvermind(
  mockEffects?: NestedPartial<Config["effects"]>,
  initialState?: NestedPartial<Config["state"]>
): OvermindMock<Config> {
  const mergedState = merge(cloneDeep(config), { state: initialState });
  return createOvermindMock(mergedState, mockEffects);
}
