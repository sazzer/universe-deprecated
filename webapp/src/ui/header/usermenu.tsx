import React from "react";
import { useTranslation } from "react-i18next";
import { Link } from "react-router-dom";
import { useOvermind } from "../../overmind";

/** Header menu for the current user actions when the user is logged in */
export const UserMenu: React.FC = () => {
  const { t } = useTranslation();
  const { state } = useOvermind();

  let username;
  const currentUser =
    state.authentication.userId !== null
      ? state.users.users[state.authentication.userId]
      : undefined;
  if (currentUser !== undefined) {
    username = currentUser.displayName;
  }

  return (
    <li className="nav-item dropdown">
      <a
        className="nav-link dropdown-toggle"
        href="/"
        id="navbarDropdown"
        role="button"
        data-toggle="dropdown"
        aria-haspopup="true"
        aria-expanded="false"
      >
        {username}
      </a>
      <div className="dropdown-menu" aria-labelledby="navbarDropdown">
        <Link to="/profile" className="dropdown-item">
          {t("header.userMenu.profile")}
        </Link>
        <div className="dropdown-divider"></div>
        <a className="dropdown-item" href="/">
          {t("header.userMenu.logout")}
        </a>
      </div>
    </li>
  );
};
