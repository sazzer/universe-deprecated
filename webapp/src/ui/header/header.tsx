import React from "react";
import { useTranslation } from "react-i18next";
import { Link } from "react-router-dom";
import { LoginLink } from "./login";
import { UserMenu } from "./usermenu";
import { useOvermind } from "../../overmind";

/**
 * Component to represent the header bar of the entire page
 */
export const Header: React.FC = () => {
  const { t } = useTranslation();
  const { state } = useOvermind();

  let userMenu;
  if (state.authentication.isLoggedIn) {
    userMenu = <UserMenu />;
  } else {
    userMenu = <LoginLink />;
  }
  return (
    <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
      <Link to="/" className="navbar-brand">
        {t("page.title")}
      </Link>
      <button
        className="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label={t("header.toggleNavigation")}
      >
        <span className="navbar-toggler-icon"></span>
      </button>

      <div className="collapse navbar-collapse" id="navbarSupportedContent">
        <ul className="navbar-nav ml-auto">{userMenu}</ul>
      </div>
    </nav>
  );
};
