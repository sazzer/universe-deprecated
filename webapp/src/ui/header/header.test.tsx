import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { Header } from "./header";
import { createTestOvermind } from "../../overmind/test";
import { Provider } from "overmind-react";
import { OvermindMock, Config } from "overmind";

describe("When not logged in", () => {
  test("It renders correctly", () => {
    const overmind = createTestOvermind();
    const { container } = render(
      <Router>
        <Provider value={overmind}>
          <Header />
        </Provider>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});

describe("When logged in", () => {
  let overmind: OvermindMock<Config>;

  beforeEach(() => {
    overmind = createTestOvermind(
      {},
      {
        users: {
          users: {
            someUserId: {
              userId: "someUserId",
              username: "username",
              email: "test@example.com",
              displayName: "Test User"
            }
          }
        },
        authentication: {
          userId: "someUserId",
          accessToken: {
            accessToken: "accessToken",
            expires: new Date().toISOString()
          },
          isLoggedIn: true
        }
      }
    );
  });

  test("It renders correctly", () => {
    const { container } = render(
      <Router>
        <Provider value={overmind}>
          <Header />
        </Provider>
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});
