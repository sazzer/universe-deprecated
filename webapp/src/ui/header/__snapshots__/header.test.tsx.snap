// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When logged in It renders correctly 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      Universe
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle Navigation"
      class="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav ml-auto"
      >
        <li
          class="nav-item dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="nav-link dropdown-toggle"
            data-toggle="dropdown"
            href="/"
            id="navbarDropdown"
            role="button"
          >
            Test User
          </a>
          <div
            aria-labelledby="navbarDropdown"
            class="dropdown-menu"
          >
            <a
              class="dropdown-item"
              href="/profile"
            >
              Profile
            </a>
            <div
              class="dropdown-divider"
            />
            <a
              class="dropdown-item"
              href="/login"
            >
              Log Out
            </a>
          </div>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`When not logged in It renders correctly 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      Universe
    </a>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle Navigation"
      class="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav ml-auto"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link"
            href="/login"
          >
            Login / Register
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
